### PR TITLE
feat(medusa, link-modules): sales channel <> product module link

### PR DESCRIPTION
**Pull request by fPolic**
**What**
- add link for product module <> sales channel service
- add migrations for the pivot table
- add list method to the sales channel service

### CODE REVIEW

File: packages/link-modules/src/definitions/product-sales-channel.ts
Comment (adrien2p) [inline packages/link-modules/src/definitions/product-sales-channel.ts]: todo: rm as well the next one that is in comment

File: packages/medusa/src/repositories/sales-channel.ts
Comment (adrien2p) [inline packages/medusa/src/repositories/sales-channel.ts]: tood: this can be a breaking change, maybe you can create a private `getFreeTextSearchResults_` and keep the `getFreeTextSearchResultsAndCount` plus provide the `getFreeTextSearchResults` which under the hood will call the `getFreeTextSearchResults_` and pass the withCount. wdyt?

**Review by adrien2p — CHANGES_REQUESTED** [review]

**Review by adrien2p — APPROVED** [review]
LGTM

**Comment by adrien2p** [comment]
Have you checked if any other constraints was link to the sales channel etc?

File: packages/medusa/src/joiner-configs/sales-channel-service.ts
Comment (carlos-r-l-rodrigues) [inline packages/medusa/src/joiner-configs/sales-channel-service.ts:30]:
```suggestion
    {
      name: ["sales_channel", "sales_channels"],
      args: { entity: "SalesChannel" },
    },
```

**Comment by olivermrbl** [comment]
> Have you checked if any other constraints was link to the sales channel etc?

Perhaps Publishable API keys?

**Comment by fPolic** [comment]
> Have you checked if any other constraints was link to the sales channel etc?

There are the following relations (if that is what you meant):
1. SalesChannel <> Cart
2. SalesChannel <> Order
3. SalesChannel <> Store
4. SalesChannel <> PublishableApiKey
5. SalesChannel <> SalesChannelLocation

I was planning on covering each domain in a separate PR for easier reviewing.

File: packages/medusa/src/joiner-configs/sales-channel-service.ts
Comment (fPolic) [reply packages/medusa/src/joiner-configs/sales-channel-service.ts:30]: hmm typescript is complaining, we are defining `name` as only string in the [alias type](https://github.com/medusajs/medusa/blob/develop/packages/types/src/joiner/index.ts#L26)?

Comment (carlos-r-l-rodrigues) [reply packages/medusa/src/joiner-configs/sales-channel-service.ts:30]: I'm sorry, this is not yet in `develop` branch. I can update that later.

**Comment by olivermrbl** [comment]
@fPolic, can I get you to add a changeset before we merge?

File: packages/medusa/src/migrations/1698056997411-product-sales-channels-link.ts
Comment (olivermrbl) [inline packages/medusa/src/migrations/1698056997411-product-sales-channels-link.ts]: question: Should we add a dedicated Sales Channel feature flag?

Comment (adrien2p) [reply packages/medusa/src/migrations/1698056997411-product-sales-channels-link.ts]: I think we can, especially that we will have followup pr's

Comment (fPolic) [reply packages/medusa/src/migrations/1698056997411-product-sales-channels-link.ts]: good suggestion, I will add

**Review by adrien2p — APPROVED** [review]
LGTM, we will need to take care to include that into the core right? if the ff is enabled, then when we list the product they must take into account the sales channel. same if the product is isolated? cc @olivermrbl should we create tickets in linear to not forget what needs to be done after

**Comment by olivermrbl** [comment]
@adrien2p I just spoke with @fPolic about this. We will tackle this before moving ahead with other joiner configs. But yes, creating tickets for the things we identify to need fixing, is a great idea

**Comment by olivermrbl** [comment]
Will need a revisit + another round of reviews when #5603 is merged

File: integration-tests/factories/simple-product-factory.ts
Comment (olivermrbl) [inline integration-tests/factories/simple-product-factory.ts]: suggestion: `isMedusaV2Enabled`to be consistent with the other PRs

File: packages/medusa/src/api/routes/admin/products/get-product.ts
Comment (riqwan) [inline packages/medusa/src/api/routes/admin/products/get-product.ts]: suggestion: I think this can go directly into the `defaultAdminProductRemoteQueryObject` object since we now have a unified FF.

File: packages/medusa/src/api/routes/admin/products/list-products.ts
Comment (riqwan) [inline packages/medusa/src/api/routes/admin/products/list-products.ts]: thought: the count below will be wrong with this approach. Maybe add a note here for future update or update the count?

File: packages/medusa/src/migrations/1698056997411-product-sales-channels-link.ts
Comment (riqwan) [inline packages/medusa/src/migrations/1698056997411-product-sales-channels-link.ts]:
```suggestion
        ALTER TABLE product_sales_channel DROP COLUMN IF EXISTS "id";
```

File: packages/medusa/src/models/sales-channel.ts
Comment (riqwan) [inline packages/medusa/src/models/sales-channel.ts]: todo: missed FF migration

**Review by riqwan — APPROVED** [review]
added some comments, but lgtm overall!

File: packages/medusa/src/models/sales-channel.ts
Comment (fPolic) [reply packages/medusa/src/models/sales-channel.ts]: good catch, fixed

File: packages/medusa/src/api/routes/admin/products/get-product.ts
Comment (fPolic) [reply packages/medusa/src/api/routes/admin/products/get-product.ts]: agree, I've just done that but didn't push a commit yet

File: packages/medusa/src/api/routes/admin/products/list-products.ts
Comment (fPolic) [reply packages/medusa/src/api/routes/admin/products/list-products.ts]: yep the current approach should be revisited, I will add a todo

**Review by carlos-r-l-rodrigues — APPROVED** [review]

File: packages/medusa/src/api/routes/admin/products/list-products.ts
Comment (olivermrbl) [inline packages/medusa/src/api/routes/admin/products/list-products.ts]: question: Can we remove this todo?

File: packages/medusa/src/api/routes/store/products/list-products.ts
Comment (olivermrbl) [inline packages/medusa/src/api/routes/store/products/list-products.ts]: comment: Same as other comment

File: packages/medusa/src/repositories/sales-channel.ts
Comment (olivermrbl) [inline packages/medusa/src/repositories/sales-channel.ts]: suggestion: isMedusaV2Enabled

File: packages/medusa/src/services/sales-channel.ts
Comment (olivermrbl) [inline packages/medusa/src/services/sales-channel.ts]: suggestion:
```suggestion
      const isMedusaV2Enabled =
```

File: packages/medusa/src/api/routes/store/products/get-product.ts
Comment (olivermrbl) [inline packages/medusa/src/api/routes/store/products/get-product.ts]: question: Missing todo?

**Review by olivermrbl — APPROVED** [review]

File: packages/medusa/src/api/routes/admin/products/list-products.ts
Comment (fPolic) [reply packages/medusa/src/api/routes/admin/products/list-products.ts]: yes, resolved

**Comment by olivermrbl** [comment]
LGTM, let's merge when 1.19 is out. Will release mid-week

**Comment by olivermrbl** [comment]
@fPolic, can I get you to resolve conflicts, so we can merge?

**Pull request merged**
